### DISCUSSION TITLE
v5.0.0.alpha.21 Why the audio event change ?

### DISCUSSION THREAD

**Discussion by RedsTom** [discord-jda/JDA#2277] [Feedback & Suggestions]
I don't really understand the point of the change concerning the audio channel events change.

> Additionally, events for voice channels of guild members have been changed slightly. Previously, we provided 3 separate voice channel update events:
> 
> - GuildVoiceJoinEvent
> - GuildVoiceLeaveEvent
> - GuildVoiceMoveEvent
> 
> Due to the confusing nature of these events, we have decided to instead only provide a single [GuildVoiceUpdateEvent](https://ci.dv8tion.net/job/JDA5/javadoc/net/dv8tion/jda/api/events/guild/voice/GuildVoiceUpdateEvent.html). This new event provides the old and new channel, which can each be null to indicate either a leave (null new channel) or join (null old channel).

This just provides more confusion : they are different events happening and must not be treated as the same in code.

**Comment by MinnDevelopment** [comment]
This provides less confusion as it's a single source of truth without ambiguity. The old events caused a lot of questions, asking why one of them didn't work, because it was using the move event instead. This was a *very* common issue and thus had to be resolved.

> they are different events happening 

They are quite literally the same event in the Discord API. It is simply an update to the `channel_id` of the voice state. You can easily handle this differentiation by checking for null.

```java
if (event.getChannelLeft() == null) {
  ... join code ...
} else if (event.getChannelJoined() == null) {
  ... left code ...
} else {
  ... move code ...
}
```

**Reply by RedsTom** [reply]
That's exactly what is problematic. A code without branching before now has to have at least one branching to check if the event we got is the right event. Moreover, to me, this is not great to null check to determine if this is one event or an other.

Maybe they are quite the same, but in this case, provide a common event and specific ones.

**Comment by dscalzi** [comment]
I think this is a solid change. It actually does simplify things since fundamentally you want to handle a change in the audio state, and need only be concerned with a single event. The only improvement could be wrapping the null/not null logic around an enum, so you could have like

```java
event.getVoiceUpdateType()

if(event.getVoiceUpdateType() == VoiceUpdateType.MOVE)
   // logic
```

This would really only serve to write more vebose code though, arguably an improvement for legibility but not needed otherwise.
